cluster-bootstrap/src/test/scala/akka/management/cluster/bootstrap/contactpoint/ClusterBootstrapDiscoveryBackoffIntegrationSpec.scala

```diff
@@ -0,0 +1,161 @@
+                discovery-method = akka.mock-dns
+
+                service-namespace = "svc.cluster.local"
+
+                stable-margin = 4 seconds
+              }
+            }
+          }
+        }
+        """.stripMargin).withFallback(ConfigFactory.load())
+    }
+
+    val systemA = ActorSystem("System", config("A"))
+    val systemB = ActorSystem("System", config("B"))
+
+    val clusterA = Cluster(systemA)
+    val clusterB = Cluster(systemB)
+
+    val bootstrapA = ClusterBootstrap(systemA)
+    val bootstrapB = ClusterBootstrap(systemB)
+
+    val baseTime = System.currentTimeMillis()
+    case class DiscoveryRequest(time: Long, attempt: Int, res: Future[_]) {
+      override def toString = s"DiscoveryRequest(${(time - baseTime).millis}, $attempt, $res)"
+    }
+    val resolveProbe = TestProbe()(systemA)
+
+    // prepare the "mock DNS" - this resolves to unreachable addresses for the first three
+    // times it is called, thus testing that discovery is called multiple times and
+    // that formation will eventually succeed once discovery returns reachable addresses
+
+    @volatile var called = 0
+    val name = "system.svc.cluster.local"
+
+    MockDiscovery.set(name, { () =>
+      called += 1
+
+      val res =
+        if (called < 4)
+          Future.failed(new Exception("Boom! Discovery failed, was rate limited for example..."))
+        else
+          Future.successful(Resolved(name,
+              List(
+                ResolvedTarget(clusterA.selfAddress.host.get, contactPointPorts.get("A")),
+                ResolvedTarget(clusterB.selfAddress.host.get, contactPointPorts.get("B"))
+              )))
+
+      resolveProbe.ref ! DiscoveryRequest(System.currentTimeMillis(), called, res)
+      res
+    })
+
+    "start listening with the http contact-points on 3 systems" in {
+      def start(system: ActorSystem, contactPointPort: Int) = {
+        import system.dispatcher
+        implicit val sys = system
+        implicit val mat = ActorMaterializer()(system)
+
+        val bootstrap = ClusterBootstrap(system)
+        val routes = new HttpClusterBootstrapRoutes(bootstrap.settings).routes
+        bootstrap.setSelfContactPoint(s"http://127.0.0.1:$contactPointPort")
+        Http().bindAndHandle(RouteResult.route2HandlerFlow(routes), "127.0.0.1", contactPointPort)
+      }
+
+      start(systemA, contactPointPorts("A"))
+      start(systemB, contactPointPorts("B"))
+    }
+
+    "poll discovery in exponentially increasing backoffs until eventually joining the returned nodes" in {
+      bootstrapA.discovery.getClass should ===(classOf[MockDiscovery])
+
+      bootstrapA.start()
+      // bootstrapB.start() // deliberately not discovering from this node, however its management port is available
+
+      val pA = TestProbe()(systemA)
+      clusterA.subscribe(pA.ref, classOf[MemberUp])
+
+      pA.expectMsgType[CurrentClusterState]
+      val up1 = pA.expectMsgType[MemberUp](45.seconds)
```

Comment:
how long does it take to run this test?

Reply:
around 16 seconds I think

Reply:
[info] - should poll discovery in exponentially increasing backoffs until eventually joining the returned nodes (20 seconds, 110 milliseconds)